dipy/align/imaffine.py

```diff
@@ -144,17 +160,61 @@ def set_affine(self, affine):
             remains unchanged. If None, then `self` represents the identity
             transformation.
         """
-        self.affine = affine
-        self.affine_inv = None
-        if self.affine is None:
+
+        if affine is None:
+            self.affine = None
+            self.affine_inv = None
             return
+
+        if not isinstance(affine, np.ndarray):
+            raise TypeError('Input is not of type ndarray.')
+
         if not np.all(np.isfinite(affine)):
-            raise AffineInversionError('Affine contains invalid elements')
+            raise AffineInvalidValuesError('Affine contains invalid elements')
+
+        # making a copy to insulate it from changes outside object
+        self.affine = affine.copy()
+
         try:
             self.affine_inv = npl.inv(affine)
         except npl.LinAlgError:
             raise AffineInversionError('Affine cannot be inverted')
 
+    def __str__(self):
+        """Printable format - relies on ndarray's implementation."""
+
+        return str(self.affine)
+
+    def __repr__(self):
+        """Reloable represenation - also relies on ndarray's implementation."""
+
+        return self.affine.__repr__()
+
+    def __format__(self, format_spec):
+        """Implementation various formatting options"""
+
+        if format_spec is None:
+            return str(self.affine)
+        elif isinstance(format_spec, str):
+            format_spec = format_spec.lower()
+            if format_spec in ['', ' ', 'f', 'full']:
+                return str(self.affine)
+            # rotation part only (initial 3x3)
+            elif format_spec in ['r', 'rotation']:
+                return str(self.affine[:-1,:-1])
+            # translation part only (4th col)
+            elif format_spec in ['t', 'translation']:
+                # notice unusual indexing to make it a column vector
+                #   i.e. rows from 0 to n-1, cols from n to n
+                return str(self.affine[:-1,-1:])
+            else:
+                allowed_formats_print_map = ['full', 'f',
+                                             'rotation', 'r',
+                                             'translation', 't']
+                raise NotImplementedError('Format {} not recognized or implemented.\n'
+                                          'Try one of {}'.format(format_spec, allowed_formats_print_map))
+
+
```

Comment:
pep8: too many blank lines

Reply:
isn't two lines between two methods standard? Atleast my Pycharm thinks so

Reply:
The best answer is here: https://www.python.org/dev/peps/pep-0008/#blank-lines

> Method definitions inside a class are surrounded by a single blank line.